adr/2025-07-02-expressions-engine.md

```diff
@@ -0,0 +1,119 @@
+
+The expression engine is currently aimed at supporting three key features: validation, conditions, and personalisation. We believe that all three of these key concepts can be supported through a single expressions framework, and that by doing so it greatly reduces the amount of technical work required, and the complexity of the system. A suitably flexible and generic expressions language should be able to grow to meet any future needs of the funding service around applying constraints or personalisation. We believe this provides a strong, simple, and extensible foundation for the platform.
+
+### Validation
+
+For validation, the expressions engine is invoked with data that the user is currently submitting. For example, for the question "How much did you spend last year?", you might want to validate that the answer is positive, so we could apply a Managed expression of "Greater than £0", which translates to an expression statement of something like `answer > 0`.
+
+Currently, validation can only target the *current* answer being submitted, but we expect in the near-to-mid-term to allow a wider set of context. For example, you might validate that the answer to this question is greater than the answer to _some previous_ question.
+
+### Conditions
+
+Conditions apply to questions and determine whether or not the user should see that question. For example, you might first ask the question "How much did you spend last year?". If the user says they spent more than £1,000,000, the form designer may wish to ask a follow-up question asking them to justify the spend. A condition such as `q_how_much_did_you_spend > 1_000_000` could be applied to this question. If the condition evaluates to `true`, then the question would be shown. Otherwise it would not.
+
+### Personalisation
+
+In order to reduce duplication and increase reliability, we anticipate the need for some questions to include informationa dynamically, based on the context of the user providing the data. For example, we might want a question to read `What role do you have in {{ organisation name }}?`. This needs to be different for each user, so that someone from Bolton Council might see "What role do you have in Bolton Council?".
```

Comment:
```suggestion
In order to reduce duplication and increase reliability, we anticipate the need for some questions to include information dynamically, based on the context of the user providing the data. For example, we might want a question to read `What role do you have in {{ organisation name }}?`. This needs to be different for each user, so that someone from Bolton Council might see "What role do you have in Bolton Council?".
```